### PR TITLE
[EMCAL-1054] Include EMCAL cell recalibrator workflow for bad channel masking

### PR DESCRIPTION
**Pull request by mfasDa**
- Chain cell-recalibrator workfow with reconstruction
  workflow to reject bad channels seen in data
- Needs to disable cell writer within reconstruction
  within reconstruction workflow and extra cell
  writer at the end of the full chain

### CODE REVIEW

**Comment by chiarazampolli** [comment]
Hello @mfasDa ,

One CI test is failing, is it related? I cannot check myself...

Chiara

**Comment by mfasDa** [comment]
Fails with
```
[0;31mDirectory "/sw/SOURCES/O2DPG-sim-tests/1.0/0/../../../O2DPG/1312-slc7_x86-64/0" is not a git repository.[0m
```
Seems it cannot clone the git repository.

**Comment by benedikt-voelkel** [comment]
Seems unrelated. It couldn't do the tests.
There is something I need to improve...

**Comment by mfasDa** [comment]
I have been trying to test the commit locally. Unfortunately I get consistently crashes in the TPC digitizer when writing the digits (see below), so I don't even reach the point where the change comes into the game. Please merge the commit nevertheless, EMCAL simulation is supposed to have bad channel masking applied at AOD level - if this is not the case then the simulation is not GOOD for EMCAL. Would be really great to have a way to run simulation only for a subset of detectors (what do I care about the TPC if I want to test an EMCAL change) - like this for testing it is a waste of time.

```
[42753:TPCDigitizer_7]: SEGMENTATION FAULT
[42753:TPCDigitizer_7]:     2   libRIO.6.28.04.so                   0x000000013f66a538: int TStreamerInfoActions::VectorLooper::WriteBasicType<int>(TBuffer&, void*, void const*, TStreamerInfoActions::TLoopConfiguration const*, TStreamerInfoActions::TConfiguration const*)
[42753:TPCDigitizer_7]:     3   libRIO.6.28.04.so                   0x000000013f51216c: TBufferFile::ApplySequence(TStreamerInfoActions::TActionSequence const&, void*, void*)
[42753:TPCDigitizer_7]:     4   libTree.6.28.04.so                  0x000000013abe3358: TBranchElement::FillLeavesCollectionMember(TBuffer&)
[42753:TPCDigitizer_7]:     5   libTree.6.28.04.so                  0x000000013abd71c0: TBranch::FillImpl(ROOT::Internal::TBranchIMTHelper*)
[42753:TPCDigitizer_7]:     6   libTree.6.28.04.so                  0x000000013abe2630: TBranchElement::FillImpl(ROOT::Internal::TBranchIMTHelper*)
[42753:TPCDigitizer_7]:     7   libTree.6.28.04.so                  0x000000013abe279c: TBranchElement::FillImpl(ROOT::Internal::TBranchIMTHelper*)
[42753:TPCDigitizer_7]:     8   o2-sim-digitizer-workflow           0x00000001047a05b4: TBranch::Fill()
[42753:TPCDigitizer_7]:     9   o2-sim-digitizer-workflow           0x000000010497aed0: o2::tpc::TPCDPLDigitizerTask::writeToROOTFile()
[42753:TPCDigitizer_7]:     10  o2-sim-digitizer-workflow           0x000000010496e1f0: o2::tpc::TPCDPLDigitizerTask::process(o2::framework::ProcessingContext&, o2::framework::DataRef const&)::'lambda'(bool)::operator()(bool) const
[42753:TPCDigitizer_7]:     11  o2-sim-digitizer-workflow           0x000000010496b52c: o2::tpc::TPCDPLDigitizerTask::process(o2::framework::ProcessingContext&, o2::framework::DataRef const&)
[42753:TPCDigitizer_7]:     12  o2-sim-digitizer-workflow           0x0000000104969e48: o2::tpc::TPCDPLDigitizerTask::run(o2::framework::ProcessingContext&)
[42753:TPCDigitizer_7]:     13  o2-sim-digitizer-workflow           0x00000001049693e4: o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)::operator()(o2::framework::ProcessingContext&) const
[42753:TPCDigitizer_7]:     14  o2-sim-digitizer-workflow           0x00000001049693a0: decltype(static_cast<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)&>(fp)(static_cast<bool&>(fp0))) std::__1::__invoke<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::ProcessingContext&>(o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)&, bool&)
[42753:TPCDigitizer_7]:     15  o2-sim-digitizer-workflow           0x0000000104969350: void std::__1::__invoke_void_return_wrapper<void, true>::__call<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::ProcessingContext&>(o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)&, o2::framework::ProcessingContext&)
[42753:TPCDigitizer_7]:     16  o2-sim-digitizer-workflow           0x00000001049692fc: std::__1::__function::__alloc_func<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&), std::__1::allocator<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)>, void (o2::framework::ProcessingContext&)>::operator()(o2::framework::ProcessingContext&)
[42753:TPCDigitizer_7]:     17  o2-sim-digitizer-workflow           0x0000000104967c38: std::__1::__function::__func<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&), std::__1::allocator<o2::framework::AlgorithmSpec o2::framework::adaptFromTask<o2::tpc::TPCDPLDigitizerTask, bool&>(bool&)::'lambda'(o2::framework::InitContext&)::operator()(o2::framework::InitContext&) const::'lambda'(o2::framework::ProcessingContext&)>, void (o2::framework::ProcessingContext&)>::operator()(o2::framework::ProcessingContext&)
[42753:TPCDigitizer_7]:     18  o2-sim-digitizer-workflow           0x000000010474d824: std::__1::__function::__value_func<void (o2::framework::ProcessingContext&)>::operator()(o2::framework::ProcessingContext&) const
[42753:TPCDigitizer_7]:     19  o2-sim-digitizer-workflow           0x000000010474d5ec: std::__1::function<void (o2::framework::ProcessingContext&)>::operator()(o2::framework::ProcessingContext&) const
[42753:TPCDigitizer_7]:     20  libO2Framework.dylib                0x000000013fd7f7c8: o2::framework::DataProcessingDevice::tryDispatchComputation(o2::framework::ServiceRegistryRef, std::__1::vector<o2::framework::DataRelayer::RecordAction, std::__1::allocator<o2::framework::DataRelayer::RecordAction>>&)::$_28::operator()(o2::framework::DataRelayer::RecordAction&)
[42753:TPCDigitizer_7]:     21  libO2Framework.dylib                0x000000013fd79584: o2::framework::DataProcessingDevice::tryDispatchComputation(o2::framework::ServiceRegistryRef, std::__1::vector<o2::framework::DataRelayer::RecordAction, std::__1::allocator<o2::framework::DataRelayer::RecordAction>>&)
[42753:TPCDigitizer_7]:     22  libO2Framework.dylib                0x000000013fd68554: o2::framework::DataProcessingDevice::doRun(o2::framework::ServiceRegistryRef)
[42753:TPCDigitizer_7]:     23  libO2Framework.dylib                0x000000013fd66288: o2::framework::run_callback(uv_work_s*)
[42753:TPCDigitizer_7]:     24  libO2Framework.dylib                0x000000013fd7661c: o2::framework::DataProcessingDevice::Run()
[42753:TPCDigitizer_7]:     25  libfairmq.1.8.0.dylib               0x000000013bf21024: fair::mq::Device::RunWrapper()
[42753:TPCDigitizer_7]:     26  libfairmq.1.8.0.dylib               0x000000013bf2bc30: std::__1::__function::__func<fair::mq::Device::Device(fair::mq::ProgOptions*, fair::mq::tools::Version)::$_2, std::__1::allocator<fair::mq::Device::Device(fair::mq::ProgOptions*, fair::mq::tools::Version)::$_2>, void (fair::mq::State)>::operator()(fair::mq::State&&)
[42753:TPCDigitizer_7]:     27  libfairmq.1.8.0.dylib               0x000000013c012848: boost::detail::function::void_function_obj_invoker1<std::__1::function<void (fair::mq::State)>, void, fair::mq::State>::invoke(boost::detail::function::function_buffer&, fair::mq::State)
[42753:TPCDigitizer_7]:     28  libfairmq.1.8.0.dylib               0x000000013c000a3c: boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, fair::mq::State>, std::__1::__list_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::__1::pair<boost::signals2::detail::slot_meta_group, boost::optional<int>>, boost::signals2::slot<void (fair::mq::State), boost::function<void (fair::mq::State)>>, boost::signals2::mutex>>, void*>, boost::signals2::detail::connection_body<std::__1::pair<boost::signals2::detail::slot_meta_group, boost::optional<int>>, boost::signals2::slot<void (fair::mq::State), boost::function<void (fair::mq::State)>>, boost::signals2::mutex>>::dereference() const
[42753:TPCDigitizer_7]:     29  libfairmq.1.8.0.dylib               0x000000013c0001e4: boost::signals2::detail::signal_impl<void (fair::mq::State), boost::signals2::optional_last_value<void>, int, std::__1::less<int>, boost::function<void (fair::mq::State)>, boost::function<void (boost::signals2::connection const&, fair::mq::State)>, boost::signals2::mutex>::operator()(fair::mq::State)
[42753:TPCDigitizer_7]:     30  libfairmq.1.8.0.dylib               0x000000013bfffc80: fair::mq::fsm::Machine_::ProcessWork()
[42753:TPCDigitizer_7]:     31  libfairmq.1.8.0.dylib               0x000000013bfff8d8: fair::mq::StateMachine::ProcessWork()
[42753:TPCDigitizer_7]:     32  libfairmq.1.8.0.dylib               0x000000013bf30668: fair::mq::DeviceRunner::Run()
[42753:TPCDigitizer_7]:     33  libO2Framework.dylib                0x000000014039b030: doChild(int, char**, o2::framework::ServiceRegistry&, o2::framework::RunningWorkflowInfo const&, o2::framework::RunningDeviceRef, o2::framework::DriverConfig const&, o2::framework::ProcessingPolicies, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, uv_loop_s*)
[42753:TPCDigitizer_7]:     34  libO2Framework.dylib                0x00000001403a3fd8: runStateMachine(std::__1::vector<o2::framework::DataProcessorSpec, std::__1::allocator<o2::framework::DataProcessorSpec>> const&, WorkflowInfo const&, std::__1::vector<o2::framework::DataProcessorInfo, std::__1::allocator<o2::framework::DataProcessorInfo>> const&, o2::framework::CommandInfo const&, o2::framework::DriverControl&, o2::framework::DriverInfo&, o2::framework::DriverConfig&, std::__1::vector<o2::framework::DeviceMetricsInfo, std::__1::allocator<o2::framework::DeviceMetricsInfo>>&, boost::program_options::variables_map&, std::__1::vector<o2::framework::ServiceSpec, std::__1::allocator<o2::framework::ServiceSpec>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>)
[42753:TPCDigitizer_7]:     35  libO2Framework.dylib                0x00000001403ba214: doMain(int, char**, std::__1::vector<o2::framework::DataProcessorSpec, std::__1::allocator<o2::framework::DataProcessorSpec>> const&, std::__1::vector<o2::framework::ChannelConfigurationPolicy, std::__1::allocator<o2::framework::ChannelConfigurationPolicy>> const&, std::__1::vector<o2::framework::CompletionPolicy, std::__1::allocator<o2::framework::CompletionPolicy>> const&, std::__1::vector<o2::framework::DispatchPolicy, std::__1::allocator<o2::framework::DispatchPolicy>> const&, std::__1::vector<o2::framework::ResourcePolicy, std::__1::allocator<o2::framework::ResourcePolicy>> const&, std::__1::vector<o2::framework::CallbacksPolicy, std::__1::allocator<o2::framework::CallbacksPolicy>> const&, std::__1::vector<o2::framework::SendingPolicy, std::__1::allocator<o2::framework::SendingPolicy>> const&, std::__1::vector<o2::framework::ConfigParamSpec, std::__1::allocator<o2::framework::ConfigParamSpec>> const&, o2::framework::ConfigContext&)
[42753:TPCDigitizer_7]:     36  o2-sim-digitizer-workflow           0x000000010472537c: mainNoCatch(int, char**)
```

**Comment by benedikt-voelkel** [comment]
Testing it, give me 10 minutes

**Comment by benedikt-voelkel** [comment]
I get this:
```bash
[ERROR] invalid workflow in o2-emcal-cell-recalibrator-workflow: No matching output found for EMC/CELLS/1 as requested by data processor "EMCALCellRecalibrator". Candidates:
223 [ERROR] -CTP/OrbitReset/0
224 [ERROR] -EMC/BADCHANNELMAP/0
225 [ERROR] -EMC/CELLS/0
226 [ERROR] -EMC/CELLSTRGR/0
```

**Comment by mfasDa** [comment]
Seems the `--run` prevents chaining of workflows, therefore the cell-recalibrator-workflow does not find the channel created by the reco workflow.

**Review by benedikt-voelkel — APPROVED** [review]

File: MC/bin/o2dpg_sim_workflow.py [hunk 1]
```diff
@@ -1050,7 +1050,11 @@ def getDigiTaskName(det):
 
    # calorimeters
    EMCRECOtask = createTask(name='emcalreco_'+str(tf), needs=[getDigiTaskName("EMC")], tf=tf, cwd=timeframeworkdir, lab=["RECO"], mem='1500')
-   EMCRECOtask['cmd'] = '${O2_ROOT}/bin/o2-emcal-reco-workflow --input-type digits --output-type cells --infile emcaldigits.root ' + getDPL_global_options(ccdbbackend=False) + putConfigValues()
+   EMCRECOtask['cmd'] = '${O2_ROOT}/bin/o2-emcal-reco-workflow --input-type digits --output-type cells --infile emcaldigits.root --disable-root-output --subspecificationOut 1 ' + putConfigValues()
+   EMCRECOtask['cmd'] += ('',' --disable-mc')[args.no_mc_labels]
+   EMCRECOtask['cmd'] += ' | ${O2_ROOT}/bin/o2-emcal-cell-recalibrator-workflow --input-subspec 1 --output-subspec 0 --no-timecalib --no-gaincalib ' + putConfigValues()
+   EMCRECOtask['cmd'] += (' --isMC','')[args.no_mc_labels]
+   EMCRECOtask['cmd'] += ' | ${O2_ROOT}/bin/o2-emcal-cell-writer-workflow --subspec 0 ' + getDPL_global_options() + putConfigValues()
```
Comment (sawenzel) [inline MC/bin/o2dpg_sim_workflow.py:1057]: In my tests I got failures due to application of unknown config-keys in the cell-writer.
Maybe we can just take away the `putConfigValues` for this step.

**Pull request merged**